### ISSUE TITLE
Generate unapply for newtype case class

### ISSUE THREAD

**Issue by oleg-py** [estatico/scala-newtype#18]
I'm starting to use more and more of this library. In particular, newtypes are quite useful for defining custom typeclass instances:

```scala
@newtype case class FirstOf[A](value: A)
object FirstOf {
   implicit def semigroup[A]: Semigroup[A] = Semigroup.instance((a, _) => a)
}
```

It would be really nice if for these cases `unapply` would also be generated. This comes in handy in chains involving foldMap, e.g.

```scala
list
  .foldMap(person => Map(person.name -> FirstOf(person.age))
  .map { case (name, FirstOf(age)) => ??? } // no need to use `.value` manually
```

**Comment by carymrobbins** [comment]
@oleg-py I need to add a benchmarks module to this library, mind sharing your JMH code? May provide a way for us to experiment with implementations to prove out zero (or near-zero) cost to get something like what you're looking for.

**Comment by joroKr21** [comment]
Also note that `unapply` will match any instance of the underlying type (unlike value classes).

**Comment by carymrobbins** [comment]
@joroKr21 Good point. However, as part of #10 I've been experimenting with different ClassTag encodings and have considered trying to just use the underlying type's ClassTag. If that happens, it may actually forbid matching invalid types (whereas now it will cause a runtime exception not caught by the compiler).

```scala
scala> :paste
@newtype case class Foo(x: Int)
object Foo {
  def unapply(foo: Foo): Option[Int] = Some(foo.x)
}

// Exiting paste mode, now interpreting.

defined type alias Foo
defined trait Foo$Types
defined object Foo

scala> ("foo": Any) match { case Foo(x) => x }
java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Integer
  at scala.runtime.BoxesRunTime.unboxToInt(BoxesRunTime.java:101)
  at Foo$Ops$newtype$.x$extension(<pastie>:14)
  at Foo$.unapply(<pastie>:16)
  ... 38 elided
```

**Comment by oleg-py** [comment]
@carymrobbins I put the project here
https://github.com/oleg-py/newtype-unapply

Note the changes: 
* Int replaced with String to exclude boxing overhead
* Generated ops class seems to crash JMH with malformed class name, so I had to make a field `private[this]` and implement unapply in terms of `asInstanceOf`

Also these things did not seem to improve anything:
* Using `Some[String]` in signature
* Using a shared global / thread-local mutable variable with `var get` and `val isEmpty = true` to save on allocations

BTW, `get` and `isEmpty` are required to be instance methods, extensions would simply not work.

@joroKr21 it's a compile error to try to match a `Int` against a pattern expecting `Foo` when `Foo` is a newtype. It's only an issue if you try to match on e.g. `Any`, which is the opposite of what I do :)

**Comment by carymrobbins** [comment]
@oleg-py Brings up a good point that it is indeed only an issue if you try to match on `Any`, which is Scalac may just handle as a special case, not enforcing exhaustiveness. Consider what happens if we match on a `String` type instead -

```scala
scala> "foo" match { case Foo(x) => x }
<console>:16: error: scrutinee is incompatible with pattern type;
 found   : Foo
    (which expands to)  Foo.Type
 required: String
       "foo" match { case Foo(x) => x }
                             ^
```

**Comment by carymrobbins** [comment]
@oleg-py Experimental idea which improves the performance of unapply -

https://github.com/oleg-py/newtype-unapply/compare/master...carymrobbins:unapply-class

The newtype unapply bytecode [posted earlier](https://github.com/estatico/scala-newtype/issues/18#issuecomment-382054540) shows ClassTag being used, which I suspected was slowing things down. So I tried returning `Some` as you did as well as returning my own `Unapply` value class. Both performed very close to the case class version.

Here are the benchmarks via `sbt clean 'jmh:run TestBenchmark'` -
```
testCaseClass                thrpt  200  402211203.871 ±  861246.654  ops/s
testManualSimpleUnapply      thrpt  200  393148121.470 ± 4173139.812  ops/s
testManualUnapplyValueClass  thrpt  200  384478663.022 ±  870967.771  ops/s
testNewTypeSimpleUnapply     thrpt  200  110200664.917 ±  283007.861  ops/s
```

the `testManual` benchmarks are for handwritten newtypes, whereas the `testNewType` one is the macro generated one which contains ClassTag instances.

This may convince me to try to get rid of the ClassTag instances. For one, I dislike that they enable type-based pattern matching on newtypes. And now this. The only use case that I know of for them is to support Array construction, but I think there's probably a better way to pull that off. For instance, we could have a `NewTypeArray` type class to support casting to and from newtypes as well as constructing arrays for newtypes. I'm thinking this is probably a better solution that the hacks we're using with ClassTag to work around the flaky Scala Array machinery.

@joroKr21 Feel free to chime in on this as I know you originally introduced the ClassTag instances.

**Comment by joroKr21** [comment]
Yes, I introduced `ClassTag` only to support `Array`s. I would be perfectly happy with removing it if we find a replacement for that feature. Maybe even an `Array` constructor generated by the macro would be enough.

As for `unapply` I missed the point of pattern matching if the scrutinee is not a supertype. But I guess it's fine if people find it convenient.
